feat: Add new Radio compound component (#1944)
Workday/canvas-kit: modules/preview-react/radio/stories/radio.stories.mdx

import {Meta} from '@storybook/addon-docs';
import {Specifications, SymbolDoc} from '@workday/canvas-kit-docs';

> import {RadioGroup2} from '@workday/canvas-kit-preview-react/radio';

Reviewer: The "2" at the end seems unnecessary.

```suggestion
import {RadioGroup} from '@workday/canvas-kit-preview-react/radio';
```

import {Basic} from './examples/Basic';
import {Alert} from './examples/Alert';
import {Error} from './examples/Error';
import {Inverse} from './examples/Inverse';
import {LabelPosition} from './examples/LabelPosition';
import {NoValue} from './examples/NoValue';
import {RefForwarding} from './examples/RefForwarding';
import {Disabled} from './examples/Disabled';
import {Required} from './examples/Required';

> <Meta title="Preview/Radio" component={RadioGroup2} />

Reviewer: Removing the "2" here as well. Also, it seems like this should go in `Preview > Inputs` in SB along with the Preview TextArea and TextInput.

```suggestion
<Meta title="Preview/Inputs/Radio" component={RadioGroup} />
```

# Canvas Kit Radio

Radio Buttons allow a user to select one value from a predefined list of 7 or fewer options.

[> Workday Design Reference](https://design.workday.com/components/inputs/radio-buttons)

## Installation

```sh
yarn add @workday/canvas-kit-react
```

## Usage

### Basic Example

`Radio` buttons are intended to be grouped together using a wrapping `RadioGroup` element with a
`name` attribute. This will ensure that all `<input type="radio" />` will share the same `name`,

> grouping them together.

Reviewer: Love all the detail in the `Basic Example` section! We chatted earlier about reorganizing this content so it reads less like a list of independent thoughts. I'd recommend grouping these sentences into paragraphs like so:

> `Radio` buttons are intended to be grouped together using a wrapping `RadioGroup` element with a `name` attribute. This will ensure that all `<input type="radio" />` will share the same `name`, grouping them together. Also, each `RadioGroup.Radio` must have a unique `value`. This is used to determine which radio has been selected. Additionally, you can provide an `initialValue` to the `RadioGroup` element so that a value is pre selected from the group. To tie it all together, provide an `onChange` to `RadioGroup` element that sets the selected value on [state](https://react.dev/learn/state-a-components-memory).
>
> `RadioGroup` should be used in tandem with [Form Field](/components/inputs/form-field/) where the `useFieldset` prop is set to `true` to meet accessibility standards. This ensures that the `label` text from `FormField` is attached to the `RadioGroup` and read out as a group for voiceover.
>
> `<ExampleCodeBlock code={Basic} />`
>
> Because the `RadioGroup` and `RadioGroup.Radio` have the correct WAI-ARIA
[properties](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/radiogroup_role#associated_wai-aria_roles_states_and_properties), keyboard navigation is supported.

The first paragraph introduces the reader to the component and tells them how to set the values of the Radio buttons and read the selected value of the RadioGroup. The second paragraph layers FormField on top of things. I opted to put the example next since I felt the two preceding paragraphs provided enough foundation. I included the bit about keyboard navigation after the example since it felt more supplementary.

I copied your sentences verbatim in my suggestion above and just reordered them. However, I think we can also remove some filler in there to tidy things up. For example:

> Additionally, you can provide an initialValue to the RadioGroup element so that a value is pre selected from the group.

...could be...

> You may provide an initialValue to the RadioGroup to pre-select a value

I'll leave the rest as an exercise for you 😄 We can cover further tweaks in our next pass.

`name` attribute. This will ensure that all `<input type="radio" />` will share the same `name`,
grouping them together. Also, each `RadioGroup.Radio` must have a unique `value`. This is used to
determine which radio has been selected. You can provide a `value` to the `RadioGroup` element so
that a value is pre selected from the group. To tie it all together, provide an `onChange` to
`RadioGroup` element that sets the selected value on
[state](https://react.dev/learn/state-a-components-memory).

`RadioGroup` should be used in tandem with [Form Field](/components/inputs/form-field/) where the
`useFieldset` prop is set to `true` to meet accessibility standards. This ensures that the `label`
text from `FormField` is attached to the `RadioGroup` and read out as a group for voiceover.

<ExampleCodeBlock code={Basic} />

Because the `RadioGroup` and `RadioGroup.Radio` have the correct WAI-ARIA
[properties](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/radiogroup_role#associated_wai-aria_roles_states_and_properties),
keyboard navigation is supported.

#### Alert

To set a `RadioGroup` in an alert state, you must set a few props on the `FormField` element

> including the `error` prop to `FormField.ErrorType.Alert`, provide `hintText` for the message

Reviewer: Let's stick with how we conventionally structure our docs for Alert and Error states for input components (for example, see [Checkbox](https://canvas.workday.com/components/inputs/checkbox#tab=web&heading=error-states) or [Text Input](https://canvas.workdaydesign.com/components/inputs/text-input#examples)):

```
### Error States

Content common to both Alert and Error (i.e., "To set a `RadioGroup` in an error state, you must set a few props on the `FormField` element
including the `error` prop..." You may need to tweak this paragraph so it can apply to both the Alert and Error states).

#### Alert

Alert-specific content

Alert example

#### Error

Error-specific content

Error example
```

The `value` prop may be omitted from `RadioGroup` to render the group with no selected radios.

<ExampleCodeBlock code={NoValue} />

### Ref Forwarding

Radio buttons supports [ref forwarding](https://reactjs.org/docs/forwarding-refs.html). It will
forward `ref` to its underlying `<input type="radio">` element.

<ExampleCodeBlock code={RefForwarding} />

### Label Position

Set the `labelPosition` prop of the wrapping Form Field to designate the position of the label
relative to the Radio Group. `labelPosition` accepts the following values:

- `FormField.LabelPosition.Top` (Default)
- `FormField.LabelPosition.Left`

<ExampleCodeBlock code={LabelPosition} />

### Required

Set the `required` prop of the wrapping `FormField` to `true` to indicate that the field is
required. Labels for required fields are suffixed by a red asterisk.

<ExampleCodeBlock code={Required} />

### Disabled

Set the `disabled` prop of the Radio Button `<RadioGroup.Radio value="gluten-free" disabled />` to
prevent users from interacting with it. Be careful not to disable a pre-selected Radio Button, this
will block keyboard access from the entire Radio Group.

<ExampleCodeBlock code={Disabled} />

> ### With React Hook Formm

Reviewer: ```suggestion
### With React Hook Form
```

Reviewer 2: should this live in `examples` section as we have for `formik` example?

Reviewer 3: So i thought about this and I'm open to suggestions. I wanted this to live as part of docs because I feel like this is a high usage example and I don't want it to get lost. I think with form elements most people will use it with some type of form library. If it doesn't make sense I can add it as an example!

It will forward `ref` to its underlying `<input type="radio">` element.

<ExampleCodeBlock code={RefForwarding} />

### Label Position

Set the `labelPosition` prop of the wrapping `FormField` to designate the position of the label
relative to the `RadioGroup`. `labelPosition` accepts the following values:

- `FormField.LabelPosition.Top` (Default)
- `FormField.LabelPosition.Left`

<ExampleCodeBlock code={LabelPosition} />

### Required

Set the `required` prop of the wrapping `FormField` to `true` to indicate that the field is
required. Labels for required fields are suffixed by a red asterisk.

<ExampleCodeBlock code={Required} />

### Disabled

Set the `disabled` prop of `RadioGroup.RadioButton` to prevent users from interacting with it. Be
careful not to disable a pre-selected radio button as this will block keyboard access to the entire
`RadioGroup`.

<ExampleCodeBlock code={Disabled} />

### Custom Radio Button

Use `RadioGroup.Label` instead of `RadioGroup.RadioButton` if you need direct access to the label
and the radio input. This will allow you to apply custom styling to the text and radio input.

<ExampleCodeBlock code={Custom} />

### Standalone Radio Button

> Use `StyledRadio` when you want a styled radio button on its own without using `RadioGroup`. You
> will need to handle behavior and accessibility.

Reviewer: Reminder to update the docs to reflect the renaming of `StyledRadio` to `StyledRadioButton`.
```suggestion
Use `StyledRadioButton` when you want a styled radio button on its own without using `RadioGroup`.
You will need to handle behavior and accessibility.
```

<ExampleCodeBlock code={Basic} />

Our example uses [React state](<(https://react.dev/learn/state-a-components-memory)>) to track the
value of the `RadioGroup`.

`RadioGroup` and `RadioGroup.Radio` support keyboard navigation through the proper use of WAI-ARIA
[properties](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/radiogroup_role#associated_wai-aria_roles_states_and_properties).

### Error States

Set the error prop of the wrapping `FormField` to `FormField.ErrorType.Alert` or
`FormField.ErrorType.Error` to set the `RadioGroup` to the alert or error state, respectively. You
will also need to set the `hintId` and `hintText` props on the `FormField` to meet accessibility
standards. You must set an `aria-describedby` attribute on the `RadioGroup` element that matches the
value of `hintId` set on the `FormField` element. These attributes ensure that the alert message is
associated to the `RadioGroup` and read out by voiceover.

#### Alert

Use the alert state when a selection is valid but there is additional information.

<ExampleCodeBlock code={Alert} />

### Error

Use the error state when the selection is no longer valid.

> <ExampleCodeBlock code={Error} />

Reviewer: A couple things on the `Error States` section:

1. `### Error` should be `#### Error`
2. Can we move the entire `Error States` section so it's the last example (right above the `Component API` section)? This would be consistent with the rest of our input component MDXes which have `Error States` at the bottom.

Reviewer 3: yep!

### Label Position

Set the `labelPosition` prop of the wrapping `FormField` to designate the position of the label
relative to the `RadioGroup`. `labelPosition` accepts the following values:

- `FormField.LabelPosition.Top` (Default)
- `FormField.LabelPosition.Left`

<ExampleCodeBlock code={LabelPosition} />

### Required

Set the `required` prop of the wrapping `FormField` to `true` to indicate that the field is
required. Labels for required fields are suffixed by a red asterisk.

<ExampleCodeBlock code={Required} />

### Disabled

> Set the `disabled` prop of the Radio Button
> `<RadioGroup.RadioButton value="gluten-free" disabled />` to prevent users from interacting with it.
> Be careful not to disable a pre-selected Radio Button, this will block keyboard access from the
> entire Radio Group.

Reviewer: Cutting out unnecessary code snippet and adding backticks.
```suggestion
Set the `disabled` prop of `RadioGroup.RadioButton` to prevent users from interacting with it. Be
careful not to disable a pre-selected radio button as this will block keyboard access to the entire
`RadioGroup`.
```

<ExampleCodeBlock code={Disabled} />

> ### Standalone Radio
> 
> Use `StyledRadio` when you want a standalone styled radio input and full control of the behavior and
> accessibility.
> 
> <ExampleCodeBlock code={StandaloneRadio} />
> 
> Use `RadioGroup.Label` and its sub components when you want more control over styling the text or
> input elements but still want some behavior handled for you.
> 
> ### Custom
> 
> Use `RadioGroup.Label` and its sub components when you want more control over styling the text or
> input elements.
> 
> <ExampleCodeBlock code={Custom} />

Reviewer: A couple things here:

1. I suggest putting the Standalone example after the Custom example since the Standalone example feels like more of a last resort.
2. Added text to the Custom example explaining that `RadioGroup.Label` is an alternative to `RadioGroup.Button`.

```suggestion
### Custom Radio Button

Use `RadioGroup.Label` instead of `RadioGroup.RadioButton` if you need direct access to the label
and the radio input. This will allow you to apply custom styling to the text and radio input.

<ExampleCodeBlock code={Custom} />

### Standalone Radio Button

Use `StyledRadio` when you want a styled radio button on its own without using `RadioGroup`. You
will need to handle behavior and accessibility.

<ExampleCodeBlock code={StandaloneRadio} />

Use `RadioGroup.Label` when you want more control styling the text and radio input but still want
some behavior handled for you.
```

### Standalone Radio Button

Use `StyledRadioButton` when you want a styled radio button on its own without using `RadioGroup`.
You will need to handle behavior and accessibility.

<ExampleCodeBlock code={StandaloneRadio} />

Use `RadioGroup.Label` when you want more control styling the text and radio input but still want
some behavior handled for you.

### Error States

Set the error prop of the wrapping `FormField` to `FormField.ErrorType.Alert` or
`FormField.ErrorType.Error` to set the `RadioGroup` to the alert or error state, respectively. You
will also need to set the `hintId` and `hintText` props on the `FormField` to meet accessibility
standards. You must set an `aria-describedby` attribute on the `RadioGroup` element that matches the
value of `hintId` set on the `FormField` element. These attributes ensure that the alert message is
associated to the `RadioGroup` and read out by voiceover.

#### Alert

Use the alert state when a selection is valid but there is additional information.

<ExampleCodeBlock code={Alert} />

#### Error

Use the error state when the selection is no longer valid.

<ExampleCodeBlock code={Error} />

### React Hook Form

> Using a form library like [React Hook Form](https://www.react-hook-form.com/) is very common.

Reviewer 3: ```suggestion
Using a form library like [React Hook Form](https://www.react-hook-form.com/) is a common use case.
```
